prototypes/modified_zonal_approach2/modified_zonal_approach2.cc

```diff
@@ -0,0 +1,317 @@
+﻿/* ---------------------------------------------------------------------
+ *
+ * Copyright (C) 2019- by the Lethe authors
+ *
+ * This file is part of the Lethe library.
+ *
+ * The LEthe library is free software; you can use it, redistribute
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE at
+ * the top level of the Lethe distribution.
+ *
+ * ---------------------------------------------------------------------
+
+ *
+ * Author: Bruno Blais, Polytechnique Montreal, 2019
+ */
+
+
+// This is a template folder for prototype executables to be created when
+// developing totally new Lethe functionnalities
+
+#include <deal.II/base/logstream.h>
+#include <deal.II/base/tensor.h>
+#include <deal.II/base/point.h>
+#include "mockup_systems.h"
+#include "write_distances.h"
+#include <deal.II/base/timer.h>
+
+
+using namespace dealii;
+
+
+template <int dim>
+class ParticleTriangleDistance {
+public:
+    ParticleTriangleDistance<dim>()
+    {    }
+
+    std::vector<double> calculate_distance(std::vector<Point<dim>> triangle, std::vector<Point<dim>> particles)
+    {
+  const double        radius = 0.25;
+  auto               &p0    = triangle[0];
+  auto               &p1    = triangle[1];
+  auto               &p2    = triangle[2];
+   std::vector<double> distances(particles.size());
+
+        const Tensor<1, dim> e0 = p1 - p0;
+        const Tensor<1, dim> e1 = p2 - p0;
+
+        const Tensor<1, dim> normal      = cross_product_3d(e0, e1);
+        const double         norm_normal = normal.norm();
+        const Tensor<1, dim> unit_normal = normal / norm_normal;
+
+        const double a   = e0.norm_square();
+        const double b   = scalar_product(e0, e1);
+        const double c   = e1.norm_square();
+
+       // Pre-allocation for speed
+       Tensor<1, dim> vector_to_plane;
+       Point<dim>     pt_in_triangle;
+
+       unsigned int k = 0;
+
+    for (auto &part : particles)
+                  {
+                    vector_to_plane         = p0 - part;
+                    double distance_squared = scalar_product(vector_to_plane, unit_normal);
+
+                    // If the particle is too far from the plane, set distance squared as an
+                    // arbitrary distance and continue
+                    if (distance_squared > (radius * radius))
+                      {
+                        distances[k] = std::sqrt(distance_squared);
+                        ++k;
+                        continue;
+                      }
+
+
+                const double d = scalar_product(e0, vector_to_plane);
+                const double e = scalar_product(e1, vector_to_plane);
+
+        auto det = (a * c - b * b);
+        auto det_inverse = 1.0 / det;
```

Comment:
`det` and `det_inverse` can be computed just after `a`, `b` and `c` since they only depend on the triangle (not the particle).